core/lib/types/src/abi.rs

```diff
@@ -315,43 +366,59 @@ impl ProposedUpgrade {
         Token::Tuple(tokens)
     }
 
-    /// Decodes `ProposedUpgrade` from a RLP token.
-    /// Returns an error if token doesn't match the `schema()`.
-    pub fn decode(token: Token) -> anyhow::Result<Self> {
+    pub(crate) fn decode(data: &[u8]) -> anyhow::Result<Self> {
+        let mut upgrade = None;
+        for (schema, token) in ProposedUpgradeSchema::decode_to_token(data) {
+            match Self::decode_from_token(schema, token) {
+                Ok(decoded) => {
+                    if let Some(upgrade) = &upgrade {
+                        anyhow::bail!("Ambiguous upgrade: can be decoded as either {decoded:?} or {upgrade:?}");
```

Comment:
I don't like this approach since there's no guarantee that this error won't trigger. In practice, the new upgrade payloads decode successfully using either the post-gateway or the post-EVM schema, and the only thing stopping the post-gateway decoding from being correct is a bogus protocol version (0). Ideally, I'd imagine that the protocol version (or maybe the schema identifier itself?) would be encoded at the known offset of the payload.

Reply:
This is indeed unfortunate. Most likely changing encoding schema in L1 contracts is not possible at this point without re-audit. As a more stable workaround you can pass `DecentralizedUpgradesEventProcessor::last_seen_protocol_version + 1` down to `decode` method as a hint to choose the schema

Reply:
It could prevent some ambiguous decodings, but not all of them, IIUC: the protocol version [may stay the same with an upgrade](https://github.com/matter-labs/zksync-era/blob/298abd2f988229a7fe5ee77de1015940d1d9ad41/core/node/eth_watch/src/event_processors/decentralized_upgrades.rs#L159), which means that an ambiguous decoding I've described above would still occur if `last_seen_protocol_version.minor == ProtocolVersionId::Version26` (since the upgrade may be either a patch bump or a v27 upgrade). Also, just for my understanding: Is it possible for an upgrade to skip a `ProtocolVersionId` (e.g., a v25 -> v27 upgrade)?

Reply:
Agree, it's not perfect either but no upgrade between v26 and v27 is planned. It's possible in theory, on practice it should only happen if we upgraded some devnet, found an issue, and want to keep the devnet live, then devnet will be upgraded this way x->x+1->x+2, and others x->x+2

Reply:
I've tried this approach with passing the min expected version, but it doesn't work straightforwardly because IIUC the watcher component may receive old upgrade events; [here's the corresponding filter](https://github.com/matter-labs/zksync-era/blob/9a5147a304cc9a5dfb705d0f17b14808b1bde052/core/node/eth_watch/src/event_processors/decentralized_upgrades.rs#L117). We most probably want to distinguish between valid old upgrades that can be skipped and upgrades that the node doesn't understand, which makes your proposal not quite straightforward. So if you don't mind, I'd leave the logic as-is.

Reply:
> no upgrade between v26 and v27 is planned
We had some discussions about v26.1